src/depends/libRumorSpreading/RumorMember.cpp

```diff
@@ -0,0 +1,282 @@
+    static std::mt19937 gen(rd());
+    std::uniform_int_distribution<int> dis(0, static_cast<int>(m_peers.size() - 1));
+    return m_peers[dis(gen)];
+}
+
+void RumorMember::increaseStatValue(StatisticKey key, double value)
+{
+    if (m_statistics.count(key) <= 0) {
+        m_statistics[key] = value;
+    }
+    else {
+        m_statistics[key] += value;
+    }
+}
+
+// CONSTRUCTORS
+RumorMember::RumorMember()
+: m_id()
+, m_networkConfig(0)
+, m_peers()
+, m_rumors()
+, m_mutex()
+, m_nextMemberCb()
+{
+}
+
+RumorMember::RumorMember(const std::unordered_set<int>& peers, int id)
+: m_id(id)
+, m_networkConfig(peers.size())
+, m_peers()
+, m_rumors()
+, m_mutex()
+, m_nextMemberCb()
+{
+    toVector(peers);
+}
+
+RumorMember::RumorMember(const std::unordered_set<int>& peers, const NextMemberCb& cb, int id)
+: m_id(id)
+  , m_networkConfig(peers.size())
+  , m_peers()
+  , m_rumors()
+  , m_mutex()
+  , m_nextMemberCb(cb)
+{
+    toVector(peers);
+}
+
+
+RumorMember::RumorMember(const std::unordered_set<int>& peers,
+                         const NetworkConfig& networkConfig,
+                         int id)
+: m_id(id)
+, m_networkConfig(networkConfig)
+, m_peers()
+, m_rumors()
+, m_mutex()
+, m_nextMemberCb()
+, m_statistics()
+{
+    assert(networkConfig.networkSize() == peers.size());
+    toVector(peers);
+}
+
+RumorMember::RumorMember(const std::unordered_set<int>& peers,
+                         const NetworkConfig& networkConfig,
+                         const NextMemberCb& cb,
+                         int id)
+: m_id(id)
+, m_networkConfig(networkConfig)
+, m_peers()
+, m_rumors()
+, m_mutex()
+, m_nextMemberCb(cb)
+, m_statistics()
+{
+    assert(networkConfig.networkSize() == peers.size());
+    toVector(peers);
+}
+
+// COPY CONSTRUCTOR
+RumorMember::RumorMember(const RumorMember& other)
+: m_id(other.m_id)
+, m_networkConfig(other.m_networkConfig)
+, m_peers(other.m_peers)
+, m_rumors(other.m_rumors)
+, m_mutex()
+, m_nextMemberCb(other.m_nextMemberCb)
+, m_statistics(other.m_statistics)
+{
+}
+
+// MOVE CONSTRUCTOR
+RumorMember::RumorMember(RumorMember&& other) noexcept
+: m_id(other.m_id)
+, m_networkConfig(other.m_networkConfig)
+, m_peers(std::move(other.m_peers))
+, m_rumors(std::move(other.m_rumors))
+, m_mutex()
+, m_nextMemberCb(std::move(other.m_nextMemberCb))
+, m_statistics(std::move(other.m_statistics))
+{
+}
+
+// PUBLIC METHODS
+bool RumorMember::addRumor(int rumorId)
+{
+    std::lock_guard<std::mutex> guard(m_mutex); // critical section
+    return m_rumors.insert(std::make_pair(rumorId, &m_networkConfig)).second;
+}
+
+std::pair<int, std::vector<Message>>
+RumorMember::receivedMessage(const Message& message, int fromPeer)
+{
+    std::lock_guard<std::mutex> guard(m_mutex); // critical section
+
+    bool isNewPeer = m_peersInCurrentRound.insert(fromPeer).second;
+    increaseStatValue(NumMessagesReceived, 1);
+
+    std::vector<Message> pullMessages;
+    if (isNewPeer && message.type() == Message::PUSH) {
```

Comment:
It took me a while to get what this code block was for. Could you add a comment here, something like “Check for any other rumors that we can pull from fromPeer”?